### PR TITLE
[HZ-502] Introduce EXPLAIN statement

### PR DESCRIPTION
**Pull request by Fly-Style**
Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

### CODE REVIEW

**Comment by Fly-Style** [comment]
@gierlachg Fixed.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/UnsupportedOperationVisitor.java [hunk 1]
```diff
@@ -266,6 +268,7 @@ public Void visit(SqlDataTypeSpec type) {
         }
     }
 
+    @SuppressWarnings("all")
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/UnsupportedOperationVisitor.java:271]: We should suppress a specific warning, not `all`. But I don't have any IDE or checkstyle warnings if I remove this, maybe it's findbugs one...

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/ExplainStatementTest.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    public void test_explainStatementOrderedScanBelowUnion() {
+        IMap<Integer, Integer> map = instance().getMap("map");
+        map.put(1, 1);
+        map.put(2, 2);
+        map.put(3, 3);
+
+        String sql = "EXPLAIN PLAN FOR SELECT * FROM map UNION ALL SELECT * FROM map ORDER BY this DESC";
+
+        createMapping("map", Integer.class, Integer.class);
+        assertRowsAnyOrder(sql, asList(
```
Comment (viliam-durina) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/ExplainStatementTest.java:106]: We should use `assertRowsOrdered`, the explain response can't be in any order.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/HazelcastSqlValidator.java [hunk 3]
```diff
@@ -133,6 +143,18 @@ public SqlNode validate(SqlNode topNode) {
             return topNode;
         }
 
+        if (topNode instanceof SqlExplainStatement) {
+            SqlExplainStatement explainStatement = (SqlExplainStatement) topNode;
+            SqlNode explicandum = explainStatement.getExplicandum();
+
+            if (explicandum instanceof SqlOrderBy) {
+                explicandum = super.performUnconditionalRewrites(explicandum, false);
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/HazelcastSqlValidator.java:151]: 1. Why do we explicitly call the superclass' impl? It's unusual except for methods calling their own super implementation. 
2. Why do we do the rewrite only for `SqlOrderBy`? Calcite does it as a part of validation for each query, so should we for EXPLAIN. Or maybe we should call the validation?

`ExplainStatementTest` passes if I do both changes above.

Comment (viliam-durina) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/HazelcastSqlValidator.java:151]: It works this way: 
```java
        if (topNode instanceof SqlExplainStatement) {
            SqlExplainStatement explainStatement = (SqlExplainStatement) topNode;
            SqlNode newExplicandum = validate(explainStatement.getExplicandum());
            explainStatement.setExplicandum(newExplicandum);
            return explainStatement;
        }
```
I was experimenting also with removing this whole `if`. But I ran into issues and I gave up. Conceptually I think it should work, but there's some small difference that breaks it...

Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/HazelcastSqlValidator.java:151]: > Why do we do the rewrite only for SqlOrderBy?

It's optimisation corner case, I guess. `SqlOrderBy` is present as AST node (or `SqlNode`), but then it become embedded as part of `SqlSelect`, and node itself is removed in `performUnconditionalRewrites()`.  As a result, `ORDER BY` is absent as operator on the next phases and also doesn't present in `SUPPORTED_KINDS`.

Explain query contains explicandum query, and  `performUnconditionalRewrites()` doesn't rewrite anything for EXPLAIN. It's a reason we do it manually.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/SqlPlanImpl.java [hunk 4]
```diff
@@ -502,6 +503,47 @@ public SqlResult execute(QueryId queryId, List<Object> arguments, long timeout)
...
+        }
+
+        @Override
+        public boolean producesRows() {
+            return true;
+        }
+
+        @Override
+        public SqlResult execute(QueryId queryId, List<Object> arguments, long timeout) {
+            SqlPlanImpl.ensureNoArguments("EXPLAIN", arguments);
```
[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/SqlPlanImpl.java:541]: I had wrong feeling about the arguments, fixed.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/ExplainStatementTest.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.singletonList;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class ExplainStatementTest extends SqlTestSupport {
```
Comment (viliam-durina) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/ExplainStatementTest.java:36]: We miss a test for:
- dynamic parameters
- unsupported stmts, e.g. `EXPLAIN SHOW TABLES`

**Review by viliam-durina — APPROVED** [review]

**Pull request merged**